README.md
# Python Casim

[![PyPI](https://img.shields.io/pypi/v/py-casim.svg)](https://pypi.org/project/py-casim/)
![PyPI - Python Version](https://img.shields.io/pypi/pyversions/py-casim)
[![Build Status](https://travis-ci.org/Sergeileduc/py-casim.svg?branch=master)](https://travis-ci.org/Sergeileduc/py-casim)
[![Documentation Status](https://readthedocs.org/projects/py-casim/badge/?version=latest)](https://py-casim.readthedocs.io/en/latest/?badge=latest)
[![codecov](https://codecov.io/gh/Sergeileduc/py-casim/branch/master/graph/badge.svg)](https://codecov.io/gh/Sergeileduc/py-casim)
[![Codacy Badge](https://app.codacy.com/project/badge/Grade/dc2020cf808b4af2b40c592e8b0cec5d)](https://www.codacy.com/manual/sergei.leduc/py-casim?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=Sergeileduc/py-casim&amp;utm_campaign=Badge_Grade)
[![Requirements Status](https://requires.io/github/Sergeileduc/py-casim/requirements.svg?branch=master)](https://requires.io/github/Sergeileduc/py-casim/requirements/?branch=master)

Python package and cli to upload image on Casimages.

## In python project

```python
from py_casim import Casim

c = Casim("my_image.jpg")

link = c.get_link()

print(link)
```

To resize an image :

```python
from py_casim import Casim

c = Casim("my_image.jpg", resize=640)

link = c.get_link()

print(link)
```

To get another url/code return :
```python
from py_casim import Casim

c = Casim("my_image.jpg", resize=640)

link = c.get_share_code(3)  # return HTML code
```

See the doc for details.

## From command line

```console
foo@bar:~$ py-casim my_image.jpg
https://nsa40.casimages.com/img/xxxxxxxxxxxxxxx.png
```

or with options :

```console
foo@bar:~$ py-casim --size 640 my_image.jpg
https://nsa40.casimages.com/img/xxxxxxxxxxxxxxx.png
```

*   Free software: MIT license
*   Documentation: <https://py-casim.readthedocs.io>.

### Features

*   upload image and get his url
*   python package (import) OR command line tool

### Credits

This package was created with Cookiecutter and the `audreyr/cookiecutter-pypackage` project template.

`Cookiecutter`: <https://github.com/audreyr/cookiecutter>

`audreyr/cookiecutter-pypackage`: <https://github.com/audreyr/cookiecutter-pypackage>


HISTORY.md
# History

## 1.0.0 (2020-07-22)

*   Stable release

## 0.2.0 (2020-02-28)

*   Better Doc

## 0.1.0 (2020-02-16)

*   Configure Travis CI
*   First release on PyPI.


tests/__init__.py
"""Unit test package for py_casim."""


tests/conftest.py
"""Pytest global fixtures and conf."""

from pathlib import Path
import pytest
from requests import Session

from py_casim.casim import Casim


@pytest.fixture
def mini_bbcode():
    """BBCODE share code."""
    bbcode = ('[URL=https://www.casimages.com/i/200217113356178313.png.html]'  # noqa: E501
              '[IMG]https://nsa40.casimages.com/img/2020/02/17/mini_200217113356178313.png[/IMG][/url]')  # noqa: E501
    return bbcode


@pytest.fixture
def response():
    """Read html sample from resp.html file."""
    # with open("tests/test_tools/resp.html", "r") as f:
    #     content = f.read_text()
    path = Path("tests/test_tools/resp.html")
    content = path.read_text()
    return content


class MockResponse:
    """Fake requests resonse."""

    def __init__(self, text, status_code):
        """Fake requests response object."""
        self.text = text
        self.status_code = status_code


@pytest.fixture(autouse=True)
def mock_session(monkeypatch, response):
    """Requests.get() mocked to return {'mock_key':'mock_response'}."""
    image_id = "200217113356178313.png"

    def mock_get(*args, **kwargs):
        url = args[1]
        if url == Casim._url:
            return MockResponse("", 200)
        elif "https://www.casimages.com/ajax/s_ano_resize.php" in url:
            return MockResponse("", 200)
        elif url == Casim._url_casi_share.format(image_id):
            return MockResponse(response, 200)
        else:
            return MockResponse("", 404)

    def mock_post(*args, **kwargs):
        url = args[1]
        if url == Casim._url_upload:
            return MockResponse(image_id, 200)

    monkeypatch.setattr(Session, "get", mock_get)
    monkeypatch.setattr(Session, "post", mock_post)


tests/test_cli.py
#!/usr/bin/env python

"""Tests for `py_casim` package."""

import pytest
from click.testing import CliRunner

from py_casim import cli


@pytest.fixture(scope='module')
def runner():
    """Cli runner to be used in tests."""
    return CliRunner()


def test_cli_help(runner):
    """Test the CLI help command."""
    help_result = runner.invoke(cli.app, ['--help'])
    assert help_result.exit_code == 0
    assert 'Show this message and exit.' in help_result.output


def test_cli_size_opt(runner):
    """Test the CLI with size opt."""
    resp = runner.invoke(cli.app, ['--size', '125', 'tests/test_casim/sample image.png'])  # noqa: E501
    assert resp.exit_code == 0
    assert resp.output.rstrip() == 'https://nsa40.casimages.com/img/2020/02/17/200217113356178313.png'  # noqa: E501


def test_cli_all_opt(runner, mini_bbcode):
    """Test the CLI with --all opt."""
    resp = runner.invoke(cli.app, ['--all', 'tests/test_casim/sample image.png'])  # noqa: E501
    assert resp.exit_code == 0
    lines = resp.output.splitlines()
    assert len(lines) == 8
    assert lines[4] == mini_bbcode


def test_cli_ode_opt(runner, mini_bbcode):
    """Test the CLI with --code opt."""
    resp = runner.invoke(cli.app, ['--code', 4, 'tests/test_casim/sample image.png'])  # noqa: E501
    assert resp.exit_code == 0
    assert resp.output.rstrip() == mini_bbcode


py_casim/casim.py
"""Code to upload an image and get his share url."""

import logging
import requests

from .tools import get_share, get_all_shares


logger = logging.getLogger(__name__)


class Casim():
    """Upload image to Casimages and get share url/code."""

    # CASIMAGES
    _url = "https://www.casimages.com/"
    _url_upload = "https://www.casimages.com/upload_ano_multi.php"
    _url_casi_share = "https://www.casimages.com/codes_ano_multi.php?img={}"
    #: Valid resize values for resize keyword argument
    resize_values = ["100", "125", "320", "640", "800", "1024", "1280", "1600"]

    _headers = {
        "Accept": "application/json",
        "Accept-Language": "fr,fr-FR;q=0.8,en-US;q=0.5,en;q=0.3",
        "Accept-Encoding": "gzip, deflate, br",
        "User-Agent": "Mozilla/5.0 (X11; Linux x86_64; rv:69.0) Gecko/20100101 Firefox/69.0",  # noqa: E501
        "Cache-Control": "no-cache",
        "Connection": "keep-alive",
        "Pragma": "no-cache",
        "X-Requested-With": "XMLHttpRequest"
        }

    def __init__(self, image, resize=None):
        """Init Casim() object with image path, and optionnal resize value."""
        self.image = image
        self.resize = resize if str(resize) in self.resize_values else None
        self.image_id = None

        self.session = requests.Session()  # Session (keep cookies)
        self.session.get(Casim._url)  # Init for cookies

        self._set_resize()

        logger.info('casim created with image: "%s" and resize: %s',
                    self.image, self.resize)

    def __repr__(self):
        return f'Casim({self.image}, resize={self.resize})'

    def _set_resize(self):
        if self.resize:
            self.url_resize = (f"https://www.casimages.com/"
                               f"ajax/s_ano_resize.php?dim={self.resize}")
            self.session.get(self.url_resize)
            logger.info('ask for resize with value %s and url : %s',
                        self.resize, self.url_resize)

    def _upload_image(self):
        """Upload image and return id."""
        with open(self.image, 'rb') as f:
            file_ = {'Filedata': ('image', f, 'image/jpg')}
            r = self.session.post(Casim._url_upload,
                                  files=file_, headers=Casim._headers)

        self.image_id = r.text  # casimages share page ID
        logger.info('upload is ok, image id is %s', self.image_id)
        return self.image_id

    def _get_share(self, index=None):
        """Get share link/code.

        Args:
            index (int, optional): sqdfsdf. Defaults to None.
            0 :     Direct link (Mail & Messenger)
            1 :     Direct link (Forum, Blog, Site)
            2 :     HTML Code Thumbnail
            3 :     HTML Code Big
            4 :     Forum BBCode Thumbnail
            5 :     Forum BBCode Big
            6 :     Source Link Thumbnail
            7 :     Source Link Big

        Returns:
            str (or list): image share url (or list of share urls)
        """
        r = self.session.get(Casim._url_casi_share.format(self.image_id))
        logger.info('get() on share page returns code : %d', r.status_code)

        return get_share(r.text, index) if index else get_all_shares(r.text)

    def get_link(self):
        """Upload image and return share link (Big source link).

        Perform same thing as get_share_code(index=7).
        """
        self._upload_image()
        return self._get_share(7)

    def get_share_code(self, index=0):
        """Get share link/code.

        Args:
            index (int, optional): Url/code choice. Defaults to 0.
            0 :     Direct link (Mail & Messenger)
            1 :     Direct link (Forum, Blog, Site)
            2 :     HTML Code Thumbnail
            3 :     HTML Code Big
            4 :     Forum BBCode Thumbnail
            5 :     Forum BBCode Big
            6 :     Source Link Thumbnail
            7 :     Source Link Big

        Returns:
            str: image share url/code
        """
        self._upload_image()
        return self._get_share(index)

    def get_all(self):
        """Get list of all links/code.

            * Direct link (Mail & Messenger)
            * Direct link (Forum, Blog, Site)
            * HTML Code Thumbnail
            * HTML Code Big
            * Forum BBCode Thumbnail
            * Forum BBCode Big
            * Source Link Thumbnail
            * Source Link Big

        Returns:
            list -- all image share codes/links
        """
        self._upload_image()
        return self._get_share()


py_casim/cli.py
"""Console script for py_casim."""
import sys
import click
from .casim import Casim


@click.command()
@click.argument('filename', type=click.Path(exists=True))
@click.option('-a', '--all', is_flag=True,
              help='Get all urls and codes (HTML/BBcode) back.')
@click.option('-s', '--size', type=click.Choice(Casim.resize_values),
              default=None,
              help='Value for image resizing. Optional.')
@click.option('-c', '--code', type=click.IntRange(0, 7),
              default=None,
              help='''Choose desired url/code.
              0 Direct link (Mail & Messenger)
              1 Direct link (Forum, Blog, Site)
              2 HTML Code Thumbnail
              3 HTML Code Big
              4 Forum BBCode Thumbnail
              5 Forum BBCode Big
              6 Source Link Thumbnail
              7 Source Link Big''')
def app(filename, size, code, all):  # pylint: disable=redefined-builtin  # noqa: D301,E501
    """Upload an image to Casimages, and get url back.

    \b
    Resizing is optionnal (125x125, 320x240, 640x480, etc...)
    Default returned url is the Source Link (Big).
    Use the option to get different url format, or use --all option.

    \b
    Examples :
        py_casim my_image.jpg

        py_casim --all my_image.jpg

        py_casim -s 640 my_image.jpg

        py_casim --size 320 --code 3 my_image.jpg
    """
    # click.echo(f"Will upload {click.format_filename(filename)} to Casimages")
    # if size:
    #     click.echo(f"Resizing with value : {size}")
    # else:
    #     click.echo("No resize")
    c = Casim(filename, resize=size)
    if all:
        res = c.get_all()
        for r in res:
            click.echo(r)
    else:
        if code:
            res = c.get_share_code(code)
            click.echo(res)
        else:
            url = c.get_link()
            click.echo(url)
    return 0


if __name__ == "__main__":
    sys.exit(app())  # pragma: no cover  # pylint: disable=no-value-for-parameter  # noqa: E501
